File: skills/product/brainstorming/scripts/md2docx.js
const fs = require("fs");
const path = require("path");
const {
  Document,
  Packer,
  Paragraph,
  TextRun,
  HeadingLevel,
  Table,
  TableRow,
  TableCell,
  WidthType,
  BorderStyle,
  AlignmentType,
  ShadingType,
} = require("docx");

const INPUT = path.join(__dirname, "..", "brainstorming-mmba-plan.md");
const OUTPUT = path.join(
  __dirname,
  "..",
  "电销系统对接MMBA工作手机方案.docx"
);

// ── helpers ──────────────────────────────────────────────

function parseInlineRuns(text, baseOpts = {}) {
  const runs = [];
  // split by **bold**, ~~strike~~, `code`
  const re = /(\*\*(.+?)\*\*)|(~~(.+?)~~)|(`([^`]+?)`)/g;
  let last = 0;
  let m;
  while ((m = re.exec(text)) !== null) {
    if (m.index > last) {
      runs.push(new TextRun({ text: text.slice(last, m.index), ...baseOpts }));
    }
    if (m[2]) {
      runs.push(new TextRun({ text: m[2], bold: true, ...baseOpts }));
    } else if (m[4]) {
      runs.push(
        new TextRun({
          text: m[4],
          strike: true,
          color: "999999",
          ...baseOpts,
        })
      );
    } else if (m[6]) {
      runs.push(
        new TextRun({
          text: m[6],
          font: "Consolas",
          size: 20,
          color: "C7254E",
          ...baseOpts,
        })
      );
    }
    last = m.index + m[0].length;
  }
  if (last < text.length) {
    runs.push(new TextRun({ text: text.slice(last), ...baseOpts }));
  }
  if (runs.length === 0) {
    runs.push(new TextRun({ text: text || "", ...baseOpts }));
  }
  return runs;
}

function makeParagraph(text, options = {}) {
  return new Paragraph({ children: parseInlineRuns(text), ...options });
}

// ── parse markdown into blocks ───────────────────────────

function parseMarkdown(md) {
  const lines = md.split(/\r?\n/);
  const blocks = [];
  let i = 0;

  while (i < lines.length) {
    const line = lines[i];

    // empty line
    if (line.trim() === "") {
      i++;
      continue;
    }

    // heading
    const hm = line.match(/^(#{1,4})\s+(.*)/);
    if (hm) {
      const level = hm[1].length; // 1-4
      blocks.push({ type: "heading", level, text: hm[2] });
      i++;
      continue;
    }

    // code block
    if (line.trim().startsWith("```")) {
      const codeLines = [];
      i++;
      while (i < lines.length && !lines[i].trim().startsWith("```")) {
        codeLines.push(lines[i]);
        i++;
      }
      i++; // skip closing ```
      blocks.push({ type: "code", text: codeLines.join("\n") });
      continue;
    }

    // table (starts with |)
    if (line.trim().startsWith("|")) {
      const tableLines = [];
      while (i < lines.length && lines[i].trim().startsWith("|")) {
        tableLines.push(lines[i]);
        i++;
      }
      // parse table
      const rows = [];
      for (const tl of tableLines) {
        const cells = tl
          .split("|")
          .map((c) => c.trim())
          .filter((c) => c !== "");
        // skip separator row (---)
        if (cells.every((c) => /^[-:]+$/.test(c))) continue;
        rows.push(cells);
      }
      if (rows.length > 0) {
        blocks.push({ type: "table", rows });
      }
      continue;
    }

    // blockquote
    if (line.trim().startsWith(">")) {
      const quoteLines = [];
      while (i < lines.length && lines[i].trim().startsWith(">")) {
        quoteLines.push(lines[i].replace(/^>\s?/, ""));
        i++;
      }
      blocks.push({ type: "quote", text: quoteLines.join("\n") });
      continue;
    }

    // unordered list
    if (/^\s*[-*]\s+/.test(line)) {
      const listItems = [];
      while (i < lines.length && /^\s*[-*]\s+/.test(lines[i])) {
        listItems.push(lines[i].replace(/^\s*[-*]\s+/, ""));
        i++;
      }
      blocks.push({ type: "list", items: listItems });
      continue;
    }

    // checkbox list
    if (/^\s*-\s+\[[ x]\]\s+/.test(line)) {
      const listItems = [];
      while (i < lines.length && /^\s*-\s+\[[ x]\]\s+/.test(lines[i])) {
        const checked = lines[i].includes("[x]");
        const text = lines[i].replace(/^\s*-\s+\[[ x]\]\s+/, "");
        listItems.push({ text, checked });
        i++;
      }
      blocks.push({ type: "checklist", items: listItems });
      continue;
    }

    // horizontal rule
    if (/^---+$/.test(line.trim())) {
      i++;
      continue;
    }

    // normal paragraph
    blocks.push({ type: "paragraph", text: line });
    i++;
  }

  return blocks;
}

// ── convert blocks to docx elements ─────────────────────

function blocksToDocx(blocks) {
  const children = [];

  for (const block of blocks) {
    switch (block.type) {
      case "heading": {
        const levelMap = {
          1: HeadingLevel.HEADING_1,
          2: HeadingLevel.HEADING_2,
          3: HeadingLevel.HEADING_3,
          4: HeadingLevel.HEADING_4,
        };
        children.push(
          new Paragraph({
            children: parseInlineRuns(block.text),
            heading: levelMap[block.level] || HeadingLevel.HEADING_4,
            spacing: { before: 240, after: 120 },
          })
        );
        break;
      }

      case "paragraph": {
        children.push(
          new Paragraph({
            children: parseInlineRuns(block.text),
            spacing: { after: 80 },
          })
        );
        break;
      }

      case "quote": {
        for (const ql of block.text.split("\n")) {
          children.push(
            new Paragraph({
              children: parseInlineRuns(ql.trim(), { italics: true, color: "555555" }),
              indent: { left: 400 },
              spacing: { after: 40 },
              border: {
                left: {
                  style: BorderStyle.SINGLE,
                  size: 6,
                  color: "BBBBBB",
                  space: 10,
                },
              },
            })
          );
        }
        break;
      }

      case "code": {
        const codeLines = block.text.split("\n");
        for (const cl of codeLines) {
          children.push(
            new Paragraph({
              children: [
                new TextRun({
                  text: cl || " ",
                  font: "Consolas",
                  size: 18,
                  color: "333333",
                }),
              ],
              spacing: { after: 0, line: 276 },
              indent: { left: 300 },
              shading: {
                type: ShadingType.SOLID,
                color: "F5F5F5",
                fill: "F5F5F5",
              },
            })
          );
        }
        // spacing after code block
        children.push(new Paragraph({ spacing: { after: 80 } }));
        break;
      }

      case "list": {
        for (const item of block.items) {
          children.push(
            new Paragraph({
              children: [
                new TextRun({ text: "•  " }),
                ...parseInlineRuns(item),
              ],
              indent: { left: 400 },
              spacing: { after: 40 },
            })
          );
        }
        break;
      }

      case "checklist": {
        for (const item of block.items) {
          const prefix = item.checked ? "☑  " : "☐  ";
          children.push(
            new Paragraph({
              children: [
                new TextRun({ text: prefix }),
                ...parseInlineRuns(item.text),
              ],
              indent: { left: 400 },
              spacing: { after: 40 },
            })
          );
        }
        break;
      }

      case "table": {
        const { rows } = block;
        if (rows.length === 0) break;
        const colCount = Math.max(...rows.map((r) => r.length));

        const tableRows = rows.map((row, ri) => {
          const cells = [];
          for (let ci = 0; ci < colCount; ci++) {
            const cellText = row[ci] || "";
            const isHeader = ri === 0;
            cells.push(
              new TableCell({
                children: [
                  new Paragraph({
                    children: parseInlineRuns(cellText, {
                      bold: isHeader,
                      size: isHeader ? 21 : 20,
                    }),
                    spacing: { after: 0 },
                  }),
                ],
                shading: isHeader
                  ? {
                      type: ShadingType.SOLID,
                      color: "E8EDF2",
                      fill: "E8EDF2",
                    }
                  : undefined,
                margins: {
                  top: 40,
                  bottom: 40,
                  left: 80,
                  right: 80,
                },
              })
            );
          }
          return new TableRow({ children: cells });
        });

        children.push(
          new Table({
            rows: tableRows,
            width: { size: 100, type: WidthType.PERCENTAGE },
          })
        );
        children.push(new Paragraph({ spacing: { after: 80 } }));
        break;
      }
    }
  }

  return children;
}

// ── main ─────────────────────────────────────────────────

async function main() {
  console.log("Reading:", INPUT);
  const md = fs.readFileSync(INPUT, "utf-8");

  console.log("Parsing markdown...");
  const blocks = parseMarkdown(md);

  console.log("Generating Word document...");
  const doc = new Document({
    styles: {
      default: {
        document: {
          run: { font: "Microsoft YaHei", size: 22 },
        },
        heading1: {
          run: { font: "Microsoft YaHei", size: 36, bold: true, color: "1F2937" },
          paragraph: { spacing: { before: 360, after: 200 } },
        },
        heading2: {
          run: { font: "Microsoft YaHei", size: 30, bold: true, color: "1F4E79" },
          paragraph: { spacing: { before: 300, after: 160 } },
        },
        heading3: {
          run: { font: "Microsoft YaHei", size: 26, bold: true, color: "2E75B6" },
          paragraph: { spacing: { before: 240, after: 120 } },
        },
        heading4: {
          run: { font: "Microsoft YaHei", size: 24, bold: true, color: "404040" },
          paragraph: { spacing: { before: 200, after: 100 } },
        },
      },
    },
    sections: [
      {
        properties: {
          page: {
            margin: { top: 1440, right: 1200, bottom: 1440, left: 1200 },
          },
        },
        children: blocksToDocx(blocks),
      },
    ],
  });

  const buffer = await Packer.toBuffer(doc);
  fs.writeFileSync(OUTPUT, buffer);
  console.log("Done! Saved to:", OUTPUT);
}

main().catch((err) => {
  console.error("Error:", err);
  process.exit(1);
});

File: skills/product/brainstorming/scripts/md2docx.py
# -*- coding: utf-8 -*-
"""Markdown to Word converter for brainstorming-mmba-plan.md"""

import re
import os
from docx import Document
from docx.shared import Pt, Cm, RGBColor, Emu
from docx.enum.text import WD_ALIGN_PARAGRAPH
from docx.enum.table import WD_TABLE_ALIGNMENT
from docx.oxml.ns import qn, nsdecls
from docx.oxml import parse_xml

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
INPUT = os.path.join(SCRIPT_DIR, "..", "brainstorming-mmba-plan.md")
OUTPUT = os.path.join(SCRIPT_DIR, "..", "电销系统对接MMBA工作手机方案.docx")


# ── inline formatting ────────────────────────────────────

def add_inline_runs(paragraph, text, base_bold=False, base_size=Pt(10.5)):
    """Parse **bold**, ~~strike~~, `code` and add runs to paragraph."""
    pattern = r"(\*\*(.+?)\*\*)|(~~(.+?)~~)|(`([^`]+?)`)"
    last = 0
    for m in re.finditer(pattern, text):
        # text before match
        if m.start() > last:
            run = paragraph.add_run(text[last : m.start()])
            run.font.size = base_size
            run.font.bold = base_bold
            run.font.name = "Microsoft YaHei"
            run._element.rPr.rFonts.set(qn("w:eastAsia"), "Microsoft YaHei")
        if m.group(2):  # bold
            run = paragraph.add_run(m.group(2))
            run.font.bold = True
            run.font.size = base_size
            run.font.name = "Microsoft YaHei"
            run._element.rPr.rFonts.set(qn("w:eastAsia"), "Microsoft YaHei")
        elif m.group(4):  # strikethrough
            run = paragraph.add_run(m.group(4))
            run.font.strike = True
            run.font.color.rgb = RGBColor(0x99, 0x99, 0x99)
            run.font.size = base_size
            run.font.name = "Microsoft YaHei"
            run._element.rPr.rFonts.set(qn("w:eastAsia"), "Microsoft YaHei")
        elif m.group(6):  # code
            run = paragraph.add_run(m.group(6))
            run.font.name = "Consolas"
            run.font.size = Pt(9)
            run.font.color.rgb = RGBColor(0xC7, 0x25, 0x4E)
        last = m.end()
    # remaining text
    if last < len(text):
        run = paragraph.add_run(text[last:])
        run.font.size = base_size
        run.font.bold = base_bold
        run.font.name = "Microsoft YaHei"
        run._element.rPr.rFonts.set(qn("w:eastAsia"), "Microsoft YaHei")
    if last == 0 and not text:
        run = paragraph.add_run("")
        run.font.size = base_size


def set_cell_shading(cell, color):
    """Set cell background color."""
    shading = parse_xml(f'<w:shd {nsdecls("w")} w:fill="{color}"/>')
    cell._tc.get_or_add_tcPr().append(shading)


def set_paragraph_font(paragraph, font_name="Microsoft YaHei", size=Pt(10.5)):
    """Set default font for a paragraph."""
    for run in paragraph.runs:
        run.font.name = font_name
        run.font.size = size
        run._element.rPr.rFonts.set(qn("w:eastAsia"), font_name)


# ── parse markdown ────────────────────────────────────────

def parse_md(filepath):
    with open(filepath, "r", encoding="utf-8") as f:
        lines = f.readlines()
    lines = [l.rstrip("\n") for l in lines]

    blocks = []
    i = 0
    while i < len(lines):
        line = lines[i]

        # empty
        if line.strip() == "":
            i += 1
            continue

        # heading
        hm = re.match(r"^(#{1,4})\s+(.*)", line)
        if hm:
            blocks.append(("heading", len(hm.group(1)), hm.group(2)))
            i += 1
            continue

        # code block
        if line.strip().startswith("```"):
            code_lines = []
            i += 1
            while i < len(lines) and not lines[i].strip().startswith("```"):
                code_lines.append(lines[i])
                i += 1
            i += 1  # skip closing
            blocks.append(("code", "\n".join(code_lines)))
            continue

        # table
        if line.strip().startswith("|"):
            table_lines = []
            while i < len(lines) and lines[i].strip().startswith("|"):
                table_lines.append(lines[i])
                i += 1
            rows = []
            for tl in table_lines:
                cells = [c.strip() for c in tl.split("|")[1:-1]]
                if cells and all(re.match(r"^[-:]+$", c) for c in cells):
                    continue
                if cells:
                    rows.append(cells)
            if rows:
                blocks.append(("table", rows))
            continue

        # blockquote
        if line.strip().startswith(">"):
            quote_lines = []
            while i < len(lines) and lines[i].strip().startswith(">"):
                quote_lines.append(re.sub(r"^>\s?", "", lines[i]))
                i += 1
            blocks.append(("quote", "\n".join(quote_lines)))
            continue

        # checkbox list
        cm = re.match(r"^\s*-\s+\[[ x]\]\s+", line)
        if cm:
            items = []
            while i < len(lines) and re.match(r"^\s*-\s+\[[ x]\]\s+", lines[i]):
                checked = "[x]" in lines[i]
                txt = re.sub(r"^\s*-\s+\[[ x]\]\s+", "", lines[i])
                items.append((txt, checked))
                i += 1
            blocks.append(("checklist", items))
            continue

        # unordered list
        lm = re.match(r"^\s*[-*]\s+", line)
        if lm:
            items = []
            while i < len(lines) and re.match(r"^\s*[-*]\s+", lines[i]):
                items.append(re.sub(r"^\s*[-*]\s+", "", lines[i]))
                i += 1
            blocks.append(("list", items))
            continue

        # hr
        if re.match(r"^---+$", line.strip()):
            i += 1
            continue

        # paragraph
        blocks.append(("paragraph", line))
        i += 1

    return blocks


# ── build docx ────────────────────────────────────────────

HEADING_COLORS = {
    1: RGBColor(0x1F, 0x29, 0x37),
    2: RGBColor(0x1F, 0x4E, 0x79),
    3: RGBColor(0x2E, 0x75, 0xB6),
    4: RGBColor(0x40, 0x40, 0x40),
}

HEADING_SIZES = {
    1: Pt(18),
    2: Pt(15),
    3: Pt(13),
    4: Pt(12),
}


def build_docx(blocks, output_path):
    doc = Document()

    # default font
    style = doc.styles["Normal"]
    style.font.name = "Microsoft YaHei"
    style.font.size = Pt(10.5)
    style._element.rPr.rFonts.set(qn("w:eastAsia"), "Microsoft YaHei")

    # page margins
    for section in doc.sections:
        section.top_margin = Cm(2.54)
        section.bottom_margin = Cm(2.54)
        section.left_margin = Cm(2.2)
        section.right_margin = Cm(2.2)

    for block in blocks:
        btype = block[0]

        if btype == "heading":
            _, level, text = block
            p = doc.add_paragraph()
            add_inline_runs(p, text, base_bold=True, base_size=HEADING_SIZES.get(level, Pt(12)))
            for run in p.runs:
                run.font.color.rgb = HEADING_COLORS.get(level, RGBColor(0, 0, 0))
                run.font.bold = True
            p.paragraph_format.space_before = Pt(16 if level <= 2 else 12)
            p.paragraph_format.space_after = Pt(8)

        elif btype == "paragraph":
            _, text = block
            p = doc.add_paragraph()
            add_inline_runs(p, text)
            p.paragraph_format.space_after = Pt(4)

        elif btype == "quote":
            _, text = block
            for ql in text.split("\n"):
                p = doc.add_paragraph()
                add_inline_runs(p, ql.strip(), base_size=Pt(10))
                for run in p.runs:
                    run.font.italic = True
                    run.font.color.rgb = RGBColor(0x55, 0x55, 0x55)
                p.paragraph_format.left_indent = Cm(0.8)
                p.paragraph_format.space_after = Pt(2)
                # left border via XML
                pPr = p._element.get_or_add_pPr()
                borders = parse_xml(
                    f'<w:pBdr {nsdecls("w")}>'
                    '  <w:left w:val="single" w:sz="6" w:space="8" w:color="BBBBBB"/>'
                    "</w:pBdr>"
                )
                pPr.append(borders)

        elif btype == "code":
            _, text = block
            for cl in text.split("\n"):
                p = doc.add_paragraph()
                run = p.add_run(cl if cl else " ")
                run.font.name = "Consolas"
                run.font.size = Pt(9)
                run.font.color.rgb = RGBColor(0x33, 0x33, 0x33)
                p.paragraph_format.space_after = Pt(0)
                p.paragraph_format.space_before = Pt(0)
                p.paragraph_format.left_indent = Cm(0.6)
                # background shading
                pPr = p._element.get_or_add_pPr()
                shading = parse_xml(f'<w:shd {nsdecls("w")} w:val="clear" w:fill="F5F5F5"/>')
                pPr.append(shading)
            # spacer
            doc.add_paragraph().paragraph_format.space_after = Pt(4)

        elif btype == "list":
            _, items = block
            for item in items:
                p = doc.add_paragraph()
                run = p.add_run("•  ")
                run.font.name = "Microsoft YaHei"
                run.font.size = Pt(10.5)
                add_inline_runs(p, item)
                p.paragraph_format.left_indent = Cm(0.8)
                p.paragraph_format.space_after = Pt(2)

        elif btype == "checklist":
            _, items = block
            for item_text, checked in items:
                p = doc.add_paragraph()
                prefix = "☑  " if checked else "☐  "
                run = p.add_run(prefix)
                run.font.name = "Microsoft YaHei"
                run.font.size = Pt(10.5)
                add_inline_runs(p, item_text)
                p.paragraph_format.left_indent = Cm(0.8)
                p.paragraph_format.space_after = Pt(2)

        elif btype == "table":
            _, rows = block
            if not rows:
                continue
            col_count = max(len(r) for r in rows)
            table = doc.add_table(rows=len(rows), cols=col_count)
            table.alignment = WD_TABLE_ALIGNMENT.CENTER
            table.style = "Table Grid"

            for ri, row in enumerate(rows):
                for ci in range(col_count):
                    cell = table.cell(ri, ci)
                    cell_text = row[ci] if ci < len(row) else ""
                    # clear default paragraph
                    cell.paragraphs[0].clear()
                    p = cell.paragraphs[0]
                    is_header = ri == 0
                    add_inline_runs(p, cell_text, base_bold=is_header, base_size=Pt(9.5))
                    p.paragraph_format.space_after = Pt(0)
                    p.paragraph_format.space_before = Pt(2)
                    if is_header:
                        set_cell_shading(cell, "E8EDF2")

            # spacer
            doc.add_paragraph().paragraph_format.space_after = Pt(4)

    doc.save(output_path)
    print(f"Done! Saved to: {output_path}")


# ── main ──────────────────────────────────────────────────

if __name__ == "__main__":
    print(f"Reading: {INPUT}")
    blocks = parse_md(INPUT)
    print(f"Parsed {len(blocks)} blocks")
    print("Generating Word document...")
    build_docx(blocks, OUTPUT)
